### PR TITLE
refactor: Reduce nesting in destination server code `write`

### PR DESCRIPTION
**Pull request by erezrokah**
#### Summary

We can check `err == io.EOF` outside of `err != nil` to make it easier to follow the code

---

Use the following steps to ensure your PR is ready to be reviewed

- [x] Read the [contribution guidelines](../blob/main/CONTRIBUTING.md) 🧑‍🎓
- [x] Run `go fmt` to format your code 🖊
- [x] Lint your changes via `golangci-lint run` 🚨 (install golangci-lint [here](https://golangci-lint.run/usage/install/#local-installation))
- [x] Update or add tests 🧪
- [x] Ensure the status checks below are successful ✅

### CODE REVIEW

**Review by hermanschaaf — APPROVED** [review]

**Review by disq — APPROVED** [review]

File: internal/servers/destinations.go [hunk 1]
```diff
@@ -86,14 +86,14 @@ func (s *DestinationServer) Write2(msg pb.Destination_Write2Server) error {
...
 	for {
 		r, err := msg.Recv()
-		if err != nil {
-			if err == io.EOF {
-				close(resources)
-				if err := eg.Wait(); err != nil {
-					return fmt.Errorf("got EOF. failed to wait for plugin: %w", err)
-				}
-				return msg.SendAndClose(&pb.Write2_Response{})
+		if err == io.EOF {
```
Comment (candiduslynx) [inline internal/servers/destinations.go:89]:
```suggestion
		if errors.Is(err, io.EOF) {
```
`errors.Is` is much more reliable as the errors are wrapped more and more often.

File: internal/servers/destinations.go [hunk 2]
```diff
@@ -86,14 +86,14 @@ func (s *DestinationServer) Write2(msg pb.Destination_Write2Server) error {
...
+			close(resources)
```
Comment (candiduslynx) [inline internal/servers/destinations.go:90]: Let's move this outside the `for` loop and do a `defer` on it?
All exit routes do it, so I see no point in several places doing it.

File: internal/servers/destinations.go [hunk 3]
```diff
@@ -86,14 +86,14 @@ func (s *DestinationServer) Write2(msg pb.Destination_Write2Server) error {
...
+			if err := eg.Wait(); err != nil {
+				return fmt.Errorf("got EOF. failed to wait for plugin: %w", err)
 			}
+			return msg.SendAndClose(&pb.Write2_Response{})
+		}
+		if err != nil {
```
Comment (candiduslynx) [inline internal/servers/destinations.go:96]: Maybe we can do this with a `switch`?
```go
switch {
	case errors.Is(err, io.EOF):
		// finish
	case err != nil:
		// error handling
	default:
		// process
}
```

**Review by candiduslynx — CHANGES_REQUESTED** [review]

File: internal/servers/destinations.go (hunk 2 above)
Comment (erezrokah) [reply internal/servers/destinations.go:90]: Maybe we can do each refactor in its own PR? That would make it easier to review and keep the changes in each PR small

Comment (candiduslynx) [reply internal/servers/destinations.go:90]: Ass this PR is about restructuring code for optimizing nesting, I think moving chan close outside would also allow to optimize the same code a bit better.

Comment (erezrokah) [reply internal/servers/destinations.go:90]: Sure it's about refactoring, but if we can split the refactoring into multiple smaller PRs, is there a reason not to do it?

**Review by candiduslynx — APPROVED** [review]

**Review by candiduslynx — APPROVED** [review]

**Comment by erezrokah** [comment]
Done in https://github.com/cloudquery/plugin-sdk/pull/529

**Pull request closed without merging**